hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/meta/TimelineBasedCkpMetadata.java

```diff
@@ -0,0 +1,162 @@
+    super.commitInstant(instant);
+    sendRefreshRequest();
+  }
+
+  @Override
+  public void abortInstant(String instant) {
+    super.abortInstant(instant);
+    sendRefreshRequest();
+  }
+
+  @Override
+  protected Stream<CkpMessage> fetchCkpMessages(Path ckpMetaPath) throws IOException {
+    // Read ckp messages from timeline server
+    Stream<CkpMessage> ckpMessageStream;
+    try {
+      List<InstantStateDTO> instantStateDTOList = executeRequestToTimelineServerWithRetry(
+          InstantStateHandler.ALL_INSTANT_STATE_URL, getRequestParams(ckpMetaPath.toString()),
+          new TypeReference<List<InstantStateDTO>>() {
+          }, RequestMethod.GET);
+      ckpMessageStream = instantStateDTOList.stream().map(c -> new CkpMessage(c.getInstant(), c.getState()));
+    } catch (Exception e) {
+      LOG.error("Failed to execute scan ckp metadata, fall back to read from file system...", e);
+      // If we failed to request timeline server, read ckp messages from file system directly.
+      ckpMessageStream = super.fetchCkpMessages(ckpMetaPath);
+    }
+    return ckpMessageStream;
+  }
+
+  private Map<String, String> getRequestParams(String dirPath) {
+    return Collections.singletonMap(InstantStateHandler.INSTANT_STATE_DIR_PATH_PARAM, dirPath);
+  }
+
+  /**
+   * Refresh the ckp messages that cached in timeline server.
+   */
+  private void sendRefreshRequest() {
+    try {
+      boolean success = executeRequestToTimelineServerWithRetry(
+          InstantStateHandler.REFRESH_INSTANT_STATE, getRequestParams(path.toString()),
+          new TypeReference<Boolean>() {
+          }, RequestMethod.POST);
+      if (!success) {
+        LOG.warn("Timeline server responses with failed refresh");
+      }
+    } catch (Exception e) {
+      // Do not propagate the exception because the server will also do auto refresh
+      LOG.error("Failed to execute refresh", e);
+    }
+  }
+
+  private <T> T executeRequestToTimelineServerWithRetry(String requestPath, Map<String, String> queryParameters,
+                                                        TypeReference reference, RequestMethod method) {
+    int retry = 5;
+    while (--retry >= 0) {
+      long start = System.currentTimeMillis();
+      try {
+        return executeRequestToTimelineServer(requestPath, queryParameters, reference, method);
+      } catch (IOException e) {
+        LOG.warn("Failed to execute ckp request (" + requestPath + ") to timeline server", e);
+      } finally {
+        LOG.info("Execute request : (" + requestPath + "), costs: " + (System.currentTimeMillis() - start) + " ms");
+      }
+    }
+    throw new HoodieException("Failed to execute ckp request (" + requestPath + ")");
+  }
+
+  private <T> T executeRequestToTimelineServer(String requestPath, Map<String, String> queryParameters,
+                                               TypeReference reference, RequestMethod method) throws IOException {
+    URIBuilder builder =
```

Comment:
Can this code be reused?

Reply:
> Can this code be reused?

Sure.  Will reuse the code in  `TimelineServerBasedWriteMarkers` and add a new Class `TimelineServerHelper`.